lib/puppet/pops/types/type_calculator.rb

```diff
@@ -313,240 +225,111 @@ def equals(left, right)
+      type = Types::PFloatType::DEFAULT
     when c == Numeric
-      type = Types::PNumericType.new()
+      type = Types::PNumericType::DEFAULT
     when c == String
-      type = Types::PStringType.new()
+      type = Types::PStringType::DEFAULT
     when c == Regexp
-      type = Types::PRegexpType.new()
+      type = Types::PRegexpType::DEFAULT
     when c == NilClass
-      type = Types::PUndefType.new()
+      type = Types::PUndefType::DEFAULT
     when c == FalseClass, c == TrueClass
-      type = Types::PBooleanType.new()
+      type = Types::PBooleanType::DEFAULT
     when c == Class
-      type = Types::PType.new()
+      type = Types::PType::DEFAULT
     when c == Array
       # Assume array of data values
-      type = Types::PArrayType.new()
-      type.element_type = Types::PDataType.new()
+      type = Types::PArrayType::DATA
     when c == Hash
       # Assume hash with scalar keys and data values
-      type = Types::PHashType.new()
-      type.key_type = Types::PScalarType.new()
-      type.element_type = Types::PDataType.new()
-    else
-      type = Types::PRuntimeType.new(:runtime => :ruby, :runtime_type_name => c.name)
+      type = Types::PHashType::DATA
+   else
+      type = Types::PRuntimeType.new(:ruby, c.name)
     end
     type
   end
 
-  # Generalizes value specific types. The given type is mutated and returned.
+  # Generalizes value specific types. The generalized type is returned.
   # @api public
-  def generalize!(o)
-    @@generalize_visitor.visit_this_0(self, o)
-    o.eAllContents.each { |x| @@generalize_visitor.visit_this_0(self, x) }
+  def generalize(o)
+    o = o.generalize if o.is_a?(Types::PAnyType)
     o
   end
 
-  def generalize_Object(o)
-    # do nothing, there is nothing to change for most types
-  end
-
-  # @return [Boolean] true if the given argument is contained in a struct element key
-  def is_struct_element_key?(o)
-    c = o.eContainer
-    if c.is_a?(Types::POptionalType)
-      o = c
-      c = c.eContainer
-    end
-    c.is_a?(Types::PStructElement) && c.key_type.equal?(o)
-  end
-  private :is_struct_element_key?
-
-  def generalize_PStringType(o)
-    # Skip generalization if the string is contained in a PStructElement key.
-    unless is_struct_element_key?(o)
-      o.values = []
-      o.size_type = nil
-    end
-  end
-
-  def generalize_PCollectionType(o)
-    # erase the size constraint from Array and Hash (if one exists, it is transformed to -Infinity - + Infinity, which is
-    # not desirable.
-    o.size_type = nil
-  end
-
-  def generalize_PFloatType(o)
-    o.to = nil
-    o.from = nil
-  end
-
-  def generalize_PIntegerType(o)
-    o.to = nil
-    o.from = nil
-  end
-
   # Answers 'what is the single common Puppet Type describing o', or if o is an Array or Hash, what is the
   # single common type of the elements (or keys and elements for a Hash).
   # @api public
   #
   def infer(o)
-    @@infer_visitor.visit_this_0(self, o)
+    case o
+      when String
```

Comment:
Suggest comment that this is an optimization for most common cases (at least I think this is why it avoids calling the visitor)